org-code-javabuilder/lib/src/main/java/org/code/javabuilder/AWSFileWriter.java

```diff
@@ -0,0 +1,40 @@
+package org.code.javabuilder;
+
+import com.amazonaws.services.s3.AmazonS3;
+import java.io.InputStream;
+import org.code.protocol.FileWriter;
+import org.code.protocol.JavabuilderException;
+
+public class AWSFileWriter implements FileWriter {
+  private final String outputBucketName;
+  private final AmazonS3 s3Client;
+  private final String javabuilderSessionId;
+  private final String getOutputURL;
+  private int writes;
+
+  // Temporary limit on writes per session until we can more fully limit usage.
+  private static final int WRITES_PER_SESSION = 2;
+
+  public AWSFileWriter(
+      AmazonS3 s3Client,
+      String outputBucketName,
+      String javabuilderSessionId,
+      String getOutputURL) {
+    this.outputBucketName = outputBucketName;
+    this.s3Client = s3Client;
+    this.javabuilderSessionId = javabuilderSessionId;
+    this.writes = 0;
+    this.getOutputURL = getOutputURL;
+  }
+
+  @Override
+  public String writeToFile(String filename, InputStream input) throws JavabuilderException {
+    if (this.writes >= WRITES_PER_SESSION) {
+      throw new UserInitiatedException(UserInitiatedExceptionKey.TOO_MANY_WRITES);
+    }
+    String filePath = this.javabuilderSessionId + "/" + filename;
+    this.s3Client.putObject(this.outputBucketName, filePath, input, null);
+    this.writes++;
```

Comment:
Do we need to worry about resetting `writes` if writing fails? Or does `s3client` handle all the retry logic for us?

Reply:
I'll double check what happens if `putObject` fails, good call out

Reply:
I added a put with retry method that tries to write 3 times if there are failures. If that fails we will throw an exception because this is a critical failure (student wouldn't be able to load audio and/or gif)

Reply:
Some retry / timeout patterns we’ve adopted in our main repository are here. Over time we may want to include this type of logic here as well https://github.com/code-dot-org/code-dot-org/blob/staging/lib/cdo/aws/s3.rb